_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+
+##### `#namespace_module_transaction`
+
+Reloading a module can fail due to any `Exception` being encountered when evaluating the new source.  If an `exception` occurs, then the original version of the Metasploit Module `Class` or `Module` should be restored.  Removing the current namespace module, generating a new namespace module, and restoring the old namespace module on exception is handled by `#namespace_module_transaction`.
+
+[![`Msf::Modules::Loader::Base#namespace_module_transaction`](/images/msf-modules-loader-base-namespace-module-transaction.png)](https://github.com/rapid7/metasploit-framework/blob/df9db42c32b572672ebd19fa4e6c7482c9876c2f/lib/msf/core/modules/loader/base.rb#L497-L538)
+
+##### `#current_module`
+
+`wrapper_module`'s constant lookup is moved to `#current_module`
+
+[![`Msf::Modules::Loader::Base#current_module`](/images/msf-modules-loader-base-current-module.png)](https://github.com/rapid7/metasploit-framework/blob/df9db42c32b572672ebd19fa4e6c7482c9876c2f/lib/msf/core/modules/loader/base.rb#L367-L386)
+
+`Enumerable#inject` works by taking the `yieldreturn` from the block and passing it as `parent` to the next iteration of the block along with the next `module_name` from `module_names`.  If any module along the path is not defined, then `break` is called, which causes `inject` to terminate early and return `nil` instead of the current value of the block or the last `parent`.
+
+## When refactoring introduces you to lexical scope
+
+`wrapper_module`'s Module creation was extracted to `#create_module_namespace` and the built-in `Module#module_eval` was replaced with `module_eval_with_lexical_scope`.
+
+[![`Msf::Module::Loader::Base#load_module` `module_eval_with_lexical_scope`](/images/msf-modules-loader-base-load-module-module-eval-with-lexical-scope.png)](https://github.com/rapid7/metasploit-framework/blob/df9db42c32b572672ebd19fa4e6c7482c9876c2f/lib/msf/core/modules/loader/base.rb#L114-L123)
+
+The reason for this, is that refactoring into new modules and classes actually introduced a bug: the `Msf` module was lost from the lexical scope.  The loss of `Msf` in the lexical scope is an issue because there are modules (like `auxiliary/dos/ssl/dtls_changecipherspec`) that didn't fully-qualify their constants.
+
+![`auxiliary/dos/ssl/dtls_changecipherspec`](/images/auxiliary-dos-ssl-dtls-changecipherspec.png)
+
+There is no top-level `Exploit::Remote::Tcp`, but there is a `Msf::Exploit::Remote::Tcp`, so how is `include Exploit::Remote::Tcp` finding `Msf::Exploit::Remote::Tcp`?
+
+### Constant look-up in Ruby
+
+This is a slight digression to explain constant lookup in Ruby.  The following rules are applied in order:
+
+1. Look at the lexically scoped constants
+2. Look at the ancestors of the current `Module` or `Class`
+
+The tricky part comes in what counts as the lexical scope.  It turns out that the lexical scope is not as simple as listing the namespace modules.
```

Comment:
The tricky part **~~comes in~~** **is** what counts as the lexical scope.  It turns out that the lexical scope is not as simple as listing the namespace modules.